include/chainbase/chainbase.hpp

```diff
@@ -48,36 +51,214 @@ namespace chainbase {
-         {
-            return std::strcmp( a, b ) < 0;
+      shared_cow_string(const shared_cow_string& other) : _data(other._data), _alloc(other._alloc) {
+         if(_data != nullptr) {
+            ++_data->reference_count;
          }
+      }
+      shared_cow_string(shared_cow_string&& other) : _data(other._data), _alloc(other._alloc) {
+         other._data = nullptr;
+      }
+      shared_cow_string& operator=(const shared_cow_string& other) {
+         *this = shared_cow_string{other};
+         return *this;
+      }
+      shared_cow_string& operator=(shared_cow_string&& other) {
+         if (this != &other) {
+            _data = other._data;
+            other._data = nullptr;
+         }
+         return *this;
+      }
+      ~shared_cow_string() {
+         dec_refcount();
+      }
+      void resize(std::size_t new_size) {
+         impl* new_data = (impl*)&*_alloc.allocate(sizeof(impl) + new_size + 1);
+         new_data->reference_count = 1;
+         new_data->size = new_size;
+         std::size_t size = this->size();
+         if (new_size < size) {
+            std::memcpy(new_data->data, _data->data, new_size);
+         } else {
+            if(size != 0) {
+               std::memcpy(new_data->data, _data->data, size);
+            }
+            std::memset(new_data->data + size, 0, new_size - size);
+         }
+         new_data->data[new_size] = '\0';
+         dec_refcount();
+         _data = new_data;
+      }
+      void assign(const char* ptr, std::size_t size) {
+         impl* new_data = (impl*)&*_alloc.allocate(sizeof(impl) + size + 1);
+         new_data->reference_count = 1;
+         new_data->size = size;
+         if(size)
+            std::memcpy(new_data->data, ptr, size);
+         new_data->data[size] = '\0';
+         if (_data) {
+            dec_refcount();
+         }
+         _data = new_data;
+      }
+      void assign(const unsigned char* ptr, std::size_t size) {
+         assign((char*)ptr, size);
+      }
+      const char * c_str() const {
+         if (_data) return _data->data;
+         else return nullptr;
+      }
+      const char * data() const { return c_str(); }
+      std::size_t size() const {
+         if (_data) return _data->size;
+         else return 0;
+      }
+      const_iterator begin() const { return c_str(); }
+      const_iterator end() const {
+         if (_data) return _data->data + _data->size;
+         else return nullptr;
+      }
+      int compare(std::size_t start, std::size_t count, const char* other, std::size_t other_size) const {
+         std::size_t cmp_len = std::min(count, other_size);
+         const char* start_ptr = data() + start;
+         int result = std::char_traits<char>::compare(start_ptr, other, cmp_len);
+         if (result != 0) return result;
+         else if (count < other_size) return -1;
+         else if(count > other_size) return 1;
+         else return 0;
+      }
```

Comment:
I doubt that it matters for performance.  It would be better to match the behavior of std::string.  I just didn't remember that std::string::compare has a range check.